gobblin-hive-registration/src/main/java/gobblin/hive/policy/HiveRegistrationPolicyBase.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+
+import gobblin.annotation.Alpha;
+import gobblin.configuration.State;
+import gobblin.hive.HivePartition;
+import gobblin.hive.spec.HiveSpec;
+import gobblin.hive.spec.SimpleHiveSpec;
+
+
+/**
+ * A base implementation of {@link HiveRegistrationPolicy}. It obtains database name from
+ * property {@link #HIVE_DATABASE_NAME} or {@link #HIVE_DATABASE_REGEX} (group 1), obtains
+ * table name from property {@link #HIVE_TABLE_NAME} and {@link #HIVE_TABLE_REGEX} (group 1),
+ * and builds a {@link SimpleHiveSpec}.
+ *
+ * @author ziliu
+ */
+@Alpha
+public abstract class HiveRegistrationPolicyBase implements HiveRegistrationPolicy {
+
+  public static final String HIVE_REGISTRATION_POLICY = "hive.registration.policy";
+
+  public static final String HIVE_DATABASE_NAME = "hive.database.name";
+  public static final String HIVE_DATABASE_REGEX = "hive.database.regex";
+  public static final String HIVE_TABLE_NAME = "hive.table.name";
+  public static final String HIVE_TABLE_REGEX = "hive.table.regex";
+
+  /**
+   * A valid db or table name should start with an alphanumeric character, and contains only
+   * alphanumeric characters and '_'.
+   */
+  private static final Pattern VALID_DB_TABLE_NAME_PATTERN_1 = Pattern.compile("[a-z0-9][a-z0-9_]*");
+
+  /**
+   * A valid db or table name should contain at least one letter or '_' (i.e., should not be numbers only).
+   */
+  private static final Pattern VALID_DB_TABLE_NAME_PATTERN_2 = Pattern.compile(".*[a-z_].*");
+
+  protected final State props;
+
+  protected HiveRegistrationPolicyBase(State props) {
+    Preconditions.checkNotNull(props);
+    this.props = props;
+  }
+
+  /**
+   * This method first tries to obtain the database name from {@link #HIVE_DATABASE_NAME}.
+   * If this property is not specified, it then tries to obtain the database name using
+   * the first group of {@link #HIVE_DATABASE_REGEX}.
+   */
+  protected String getDatabaseName(Path path) {
+    return getDatabaseOrTableName(path, HIVE_DATABASE_NAME, HIVE_DATABASE_REGEX);
+  }
+
+  /**
+   * This method first tries to obtain the database name from {@link #HIVE_TABLE_NAME}.
+   * If this property is not specified, it then tries to obtain the database name using
+   * the first group of {@link #HIVE_TABLE_REGEX}.
+   */
+  protected String getTableName(Path path) {
+    return getDatabaseOrTableName(path, HIVE_TABLE_NAME, HIVE_TABLE_REGEX);
+  }
+
+  protected String getDatabaseOrTableName(Path path, String nameKey, String regexKey) {
+    String name = null;
+    if (this.props.contains(nameKey)) {
+      name = this.props.getProp(nameKey);
+    } else if (this.props.contains(regexKey)) {
+      String regex = this.props.getProp(regexKey);
+      name = Pattern.compile(regex).matcher(path.toString()).group();
```

Comment:
The `regex` here is only used for matching the string representation of `path`. But the `HIVE_DATABASE_REGEX`/ `HIVE_DATABASE_REGEX` is very generic naming. Can we rename it?